waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/util/ExpiringMemoizingSupplier.java

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright (C) 2016-2019 Expedia, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.waggledance.util;
+
+import com.google.common.base.Preconditions;
+
+import java.io.Serializable;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Adapted from
+ *
+ * @see com.google.common.base.Suppliers.ExpiringMemoizingSupplier
+ * to support invalidation
+ */
+
+public interface ExpiringMemoizingSupplier<T> extends Supplier<T>, Serializable {
```

Comment:
Not really happy that we have to have this class in here I think I prefer the cache in that case, why we can't use the ExpriringMemoizingSupplier from Guava?
Really like the java8 facelift otherwise :)

Reply:
I think, cache should be used for caching, not for memoization. Guava memoizing supplier doesn't support invalidation.

Reply:
I'd also prefer to use the cache rather than introducing a new class with a whole bunch of hand-crafted functionality (that isn't tested) that we need to maintain. Even if a cache isn't the best fit at least its common library code in Guava that has been battle-hardened and use in production all over the show.

Reply:
How about I write a test for it ?

Reply:
The problem is we then have to maintain this forever so I'd rather not. If this was fixing a bug or adding some new feature it would be different story but purely for an "idiomatic" change I don't think it's worth the burden on the maintainers of the project in future.

Reply:
I see your reason however I don't see any trouble in maintaining a trivial utility class. If is of such importance for you, I shall remove all the commits and leave only those that you find acceptable.

Reply:
Yes please, thanks!